src/main/java/org/folio/services/invoice/InvoiceFiscalYearsService.java

```diff
@@ -0,0 +1,121 @@
+package org.folio.services.invoice;
+
+import io.vertx.core.Future;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.folio.InvoiceWorkflowDataHolderBuilder;
+import org.folio.invoices.rest.exceptions.HttpException;
+import org.folio.models.InvoiceWorkflowDataHolder;
+import org.folio.rest.acq.model.finance.Budget;
+import org.folio.rest.core.models.RequestContext;
+import org.folio.rest.jaxrs.model.Error;
+import org.folio.rest.jaxrs.model.FiscalYear;
+import org.folio.rest.jaxrs.model.FiscalYearCollection;
+import org.folio.rest.jaxrs.model.Invoice;
+import org.folio.rest.jaxrs.model.InvoiceLine;
+import org.folio.rest.jaxrs.model.Parameter;
+import org.folio.services.finance.budget.BudgetService;
+import org.folio.services.finance.fiscalyear.FiscalYearService;
+import org.springframework.beans.factory.annotation.Autowired;
+
+import java.time.Instant;
+import java.time.LocalDate;
+import java.time.ZoneOffset;
+import java.util.List;
+import java.util.Map;
+
+import static java.util.function.Function.identity;
+import static java.util.stream.Collectors.joining;
+import static java.util.stream.Collectors.toList;
+import static java.util.stream.Collectors.toMap;
+import static org.folio.invoices.utils.ErrorCodes.COULD_NOT_FIND_VALID_FISCAL_YEAR;
+import static org.folio.invoices.utils.ErrorCodes.MORE_THAN_ONE_FISCAL_YEAR_SERIES;
+import static org.folio.invoices.utils.HelperUtils.convertIdsToCqlQuery;
+
+public class InvoiceFiscalYearsService {
+  private static final Logger log = LogManager.getLogger();
+
+  @Autowired
+  private InvoiceWorkflowDataHolderBuilder holderBuilder;
+  @Autowired
+  private BudgetService budgetService;
+  @Autowired
+  private FiscalYearService fiscalYearService;
+
+  public InvoiceFiscalYearsService(InvoiceWorkflowDataHolderBuilder holderBuilder, BudgetService budgetService,
+      FiscalYearService fiscalYearService) {
+    this.holderBuilder = holderBuilder;
+    this.budgetService = budgetService;
+    this.fiscalYearService = fiscalYearService;
+  }
+
+  public Future<FiscalYearCollection> getFiscalYearsByInvoiceAndLines(Invoice invoice, List<InvoiceLine> lines,
+      RequestContext requestContext) {
+    List<InvoiceWorkflowDataHolder> dataHolders = holderBuilder.buildHoldersSkeleton(lines, invoice);
+    List<String> fundIds = dataHolders.stream().map(InvoiceWorkflowDataHolder::getFundId).distinct().collect(toList());
+    if (fundIds.isEmpty())
```

Comment:
please add warn log to conditional statements, basically from all new features we should apply recommendations regarding logging

Reply:
okay got it